syncapi/sync/request.go

```diff
@@ -48,44 +41,50 @@ func newSyncRequest(req *http.Request, device userapi.Device, syncDB storage.Dat
 		var err error
 		since, err = types.NewStreamTokenFromString(sinceStr)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("types.NewStreamTokenFromString: %w", err)
 		}
 	}
-	timelineLimit := DefaultTimelineLimit
+
+	var f *gomatrixserverlib.Filter
 	// TODO: read from stored filters too
 	filterQuery := req.URL.Query().Get("filter")
 	if filterQuery != "" {
 		if filterQuery[0] == '{' {
 			// attempt to parse the timeline limit at least
-			var f filter
-			err := json.Unmarshal([]byte(filterQuery), &f)
-			if err == nil && f.Room.Timeline.Limit != nil {
-				timelineLimit = *f.Room.Timeline.Limit
+			if err := json.Unmarshal([]byte(filterQuery), &f); err != nil {
+				util.GetLogger(req.Context()).WithError(err).Error("json.Unmarshal failed")
+				return nil, fmt.Errorf("json.Unmarshal: %w", err)
 			}
 		} else {
 			// attempt to load the filter ID
 			localpart, _, err := gomatrixserverlib.SplitID('@', device.UserID)
 			if err != nil {
 				util.GetLogger(req.Context()).WithError(err).Error("gomatrixserverlib.SplitID failed")
-				return nil, err
+				return nil, fmt.Errorf("gomatrixserverlib.SplitID: %w", err)
 			}
-			f, err := syncDB.GetFilter(req.Context(), localpart, filterQuery)
-			if err == nil {
-				timelineLimit = f.Room.Timeline.Limit
+			f, err = syncDB.GetFilter(req.Context(), localpart, filterQuery)
+			if err != nil {
+				util.GetLogger(req.Context()).WithError(err).Error("syncDB.GetFilter failed")
+				return nil, fmt.Errorf("syncDB.GetFilter: %w", err)
 			}
 		}
 	}
 
-	filter := gomatrixserverlib.DefaultEventFilter()
-	filter.Limit = timelineLimit
-	// TODO: Additional query params: set_presence, filter
+	filter := gomatrixserverlib.DefaultFilter()
+	if f != nil {
+		filter = *f
+	}
+	limit := filter.Room.Timeline.Limit
+	if limit == 0 {
+		filter.Room.Timeline.Limit = defaultTimelineLimit
+	}
```

Comment:
hmmm, is there a way we can merge two filters into one. I'd like to merge the base with the newly parsed one.

Maybe there can be some sort of JSON parsing default in `gomatrixserverlib`

Reply:
@neilalexander Any insight?